src/librustc_arena/lib.rs

```diff
@@ -371,21 +364,27 @@ impl DroplessArena {
     }
 
     #[inline]
-    pub fn alloc_raw(&self, bytes: usize, align: usize) -> &mut [u8] {
-        unsafe {
-            assert!(bytes != 0);
-
-            self.align(align);
+    fn alloc_raw_fast(&self, bytes: usize, align: usize) -> Option<&mut [u8]> {
+        let ptr = self.ptr.get() as usize;
+        let end = self.end.get() as usize;
+        let aligned = ptr.checked_add(align - 1)? & !(align - 1);
+        let new_ptr = aligned.checked_add(bytes)?;
+        if new_ptr <= end {
+            self.ptr.set(new_ptr as *mut u8);
+            unsafe { Some(slice::from_raw_parts_mut(aligned as *mut u8, bytes)) }
+        } else {
+            None
+        }
+    }
 
-            let future_end = intrinsics::arith_offset(self.ptr.get(), bytes as isize);
-            if (future_end as *mut u8) > self.end.get() {
-                self.grow(bytes);
+    #[inline]
+    pub fn alloc_raw(&self, bytes: usize, align: usize) -> &mut [u8] {
+        assert!(bytes != 0);
+        loop {
+            if let Some(a) = self.alloc_raw_fast(bytes, align) {
+                break a;
             }
```

Comment:
What happens here on overflow? It'll call `self.grow(bytes)`. Then what will happen -- panic, success? I think the possibilities need to be clearer here. In particular, `alloc_raw_fast()` returns `None` in two cases: "overflow", and "need to grow". I feel like those two cases need to be distinguished.

Reply:
From perspective of code that tried to use the current chunk the situation is the same: there is no capacity left and arena needs to grow. Overflow does not necessarily preclude later success if say the chunk had merely a high address, requested capacity was large, and the calculation overflowed.

The code then delegates the responsibility to handle the situation to `grow` and subsequently to `RawVec`. Under assumption that grow doesn't hit `checked_mul(2).unwrap()` (this could be saturated to `isize::MAX` to be a little bit friendlier to 32-bit systems), the allocation is performed by `RawVec` which panics when requested capacity exceeds `isize::MAX` (on 32-bit systems) and aborts on OOM.

Reply:
Thanks for the explanation. Could you put that information into a comment? :)

Reply:
Now with additional comments.

Reply:
Thanks for that. Sorry to nitpick this further, but you've added a comment explaining the high-level behaviour, which is fairly obvious, while not adding a comment distinguishing the overflow vs. need-to-grow cases, which is less obvious.

Can you add a comment inside the function explaining those two cases, much like your comment above? I specifically would like to see explanation of the "high address + large capacity" case, in particular. Thanks.

Reply:
I added a *different* comment explaining the abstract check we want to perform.
In my view the overflow is the same as the need-to-grow case, which is why I
didn't want to comment about the distinction. I just don't see any.